src/main/js/main_html5.js

```diff
@@ -485,50 +486,64 @@ require("../../../html5-common/js/utils/environment.js");
      * @param {object} closedCaptions The closedCaptions object
      * @param {object} params The params to set with closed captions
      */
-    this.setClosedCaptions = function(language, closedCaptions, params) {
+    this.setClosedCaptions = _.bind(function(language, closedCaptions, params) {
+      //Remove and disable current captions before setting new ones.
       $(_video).find('.' + TRACK_CLASS).remove();
+      this.setClosedCaptionsMode(OO.CONSTANTS.CLOSED_CAPTIONS.DISABLED);
       if (language == null) return;
 
       var captionMode = (params && params.mode) || OO.CONSTANTS.CLOSED_CAPTIONS.SHOWING;
 
-      // The textTrack added by QuickTime will not be removed by removing track element
-      // But the textTrack that we added by adding track element will be removed by removing track element.
-      // This first check is to check for live CC
-      if (OO.isSafari && _video.textTracks.length !== 0 && language == "CC") {
-        for (var i = 0; i < _video.textTracks.length; i++) {
-          if (_video.textTracks[i].kind === "captions") {
-            _video.textTracks[i].mode = captionMode;
-            _video.textTracks[i].oncuechange = onClosedCaptionCueChange;
-          } else {
-           _video.textTracks[i].mode = OO.CONSTANTS.CLOSED_CAPTIONS.DISABLED;
+      //Add the new closed captions if they are valid.
+      var captionsFormat = "closed_captions_vtt";
+      if (closedCaptions && closedCaptions[captionsFormat]) {
```

Comment:
One more question.  It is legit to call setClosedCaptions with null "closedCaptions" object.  In this case this function will do nothing unless earlier call of "checkForClosedCaptions" revealed this data and populated "availableClosedCaptions" object.  This is fine as long as  checkForClosedCaptions is called more than once, but it looks like you are checking for CCs inside this function only once, on firstPlay, when loading started.  But captions data can show up much later - in case of live CCs, for example.  So it may happen that live CC data can be missed with this logic.  I think it's also possible that in-manifest CC info for VOD assets may show up later.  I think you should be calling "checkForClosedCaptions" more frequently and you may want to remove this check for firstPlay from this function.